NExtends/Collections/ObjectPropertiesList.cs

```diff
@@ -0,0 +1,362 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Linq.Expressions;
+
+namespace NExtends.Collections
+{
+	/// <summary>
+	/// <para>
+	/// Helper class for getting and setting class properties via an indexable list.
+	/// If no class attributes are specified for class <typeparamref name="K"/>, all properties will be indexed in order of definition.
+	/// </para>
+	/// 
+	/// <para>
+	/// Use custom attributes <see cref="ObjectPropertiesListProperties"/> and <see cref="ObjectPropertiesListNumberedProperties"/>
+	/// if you want to change order or only use selected properties. <seealso cref="IObjectPropertiesListParams"/>.
+	/// </para>
+	/// 
+	/// <para>
+	/// See example section for usage example.
+	/// </para>
+	/// </summary>
+	/// 
+	/// <typeparam name="K">Type of class to map properties</typeparam>
+	/// <typeparam name="T">Type of properties</typeparam>
+	/// 
+	///<example>
+	///<code>
+	///
+	///public class ObjectPropertiesListTest
+	///{		
+	///	public void ObjectFieldArrayTest1()
+	///	{
+	///		var t = new Test();
+	///		var i = new ObjectPropertiesList&lt;Test, string&gt;(t);
+	///
+	///		i[0] = "test";
+	///		i[1] = "test0";
+	///		i[2] = "test1";
+	///		i[3] = "test2";
+	///
+	///		Assert.IsTrue(t.test == "test");
+	///		Assert.IsTrue(t.field0 == "test0");
+	///		Assert.IsTrue(t.field1 == "test1");
+	///		Assert.IsTrue(t.field2 == "test2");
+	///	}
+	///}
+	///
+	///[ObjectPropertiesListProperties(0, "test")]
+	///[ObjectPropertiesListNumberedProperties(1, "field", 3, 0)]
+	///class Test
+	///{
+	///		public string test { get; set; }
+	///		public string field0 { get; set; }
+	///		public string field1 { get; set; }
+	///		public string field2 { get; set; }
+	///}
+	///</code>
+	///</example>
+	public class ObjectPropertiesList<K, T> : IList<T>
+	{
+		public static IEnumerable<string> IndexableProperties;
+		public static Func<K, T>[] Getters;
+		public static Action<K, T>[] Setters;
+		public static string[] PropertyNames;
+
+		K ObjectToIndex;
+
+		static ObjectPropertiesList()
+		{
+			var getters = new List<Func<K, T>>();
+			var setters = new List<Action<K, T>>();
+			var propertyNames = new List<string>();
+
+			var Attributes = typeof(K).GetTypeInfo().GetCustomAttributes(typeof(IObjectPropertiesListParams), true).OfType<IObjectPropertiesListParams>();
+
+			if (Attributes.Count() == 0)
+			{
+				//in case we don't have any specified attributes we use all public properties
+				IndexableProperties = typeof(K).GetProperties().Where(p => p.PropertyType == typeof(T)).Select(p => p.Name).ToList();
+			}
+			else
+			{
+				IndexableProperties = Attributes.OrderBy(a => a.Order).SelectMany(a => a.IndexableProperties).ToList();
+			}
+
+			foreach (var pName in IndexableProperties)
+			{
+				var pInfo = typeof(K).GetProperty(pName);
+
+				if (pInfo == null)
+					throw new Exception(string.Format("Class {0} doesn't have a property named {1}", typeof(K).Name, pName));
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') Remove this 'throw' statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS3877)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') 'System.Exception' should not be thrown by user code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=csharpsquid%3AS112)